WebKit 26 fails driver-integration-tests-webkit (cookies.cy.js)

### Current behavior

If an attempt is made to update [playwright-webkit](https://www.npmjs.com/package/playwright-webkit) to the latest version in Cypress source code, it fails the Cypress test spec

- [packages/driver/cypress/e2e/commands/cookies.cy.js](https://github.com/cypress-io/cypress/blob/develop/packages/driver/cypress/e2e/commands/cookies.cy.js)

which is part of `driver-integration-tests-webkit`.

This is reproducible on Ubuntu with versions of [playwright-webkit](https://www.npmjs.com/package/playwright-webkit) starting with `1.37.0` and extending to the current version (with WebKit ~~`17.4`~~ ~~`18.0`~~ `26.x`).

Cypress is currently configured to use [playwright-webkit@1.24.2](https://github.com/microsoft/playwright/releases/tag/v1.24.2) released in July 2022:

https://github.com/cypress-io/cypress/blob/d8884e0a342e92a41fb3e187a807ae7babab3cc7/package.json#L189

### Desired behavior

WebKit should pass all tests from `driver-integration-tests-webkit` when using the latest version of `playwright-webkit`.

- Qualify if this is due to bug(s) in Playwright/WebKit or Cypress
- Depending on outcome, report bugs to [Playwright](https://github.com/microsoft/playwright/issues) and / or fix bugs in Cypress

### Test code to reproduce

```shell
npm install yarn@latest -g
git clone https://github.com/cypress-io/cypress
cd cypress
yarn add playwright-webkit@latest -D -W
yarn playwright install-deps webkit # Linux only
```

then

```shell
yarn workspace @packages/driver cypress:run --browser webkit --spec cypress/e2e/commands/cookies.cy.js
```

or

```shell
yarn workspace @packages/driver cypress:open --e2e --browser webkit
```

and select `cypress/e2e/commands/cookies.cy.js` to run in UI.

### Playwright Version

First reported on [playwright-webkit@1.45.3](https://github.com/microsoft/playwright/releases/tag/v1.45.3)
Continues to be reproducible on [playwright-webkit@1.55.0](https://github.com/microsoft/playwright/releases/tag/v1.55.0)

### Cypress Version

First reported on `13.13.2`
Continues to be reproducible on `15.0.0`

### Node version

~~`v18.17.1`~~ `v20.18.1` and `v22.18.0`

### Operating System

~~Ubuntu `22.04.4` LTS~~
Ubuntu `22.04.5` LTS and
Ubuntu `24.04.3` LTS

### Debug Logs

See failures in CircleCI at
https://app.circleci.com/pipelines/github/cypress-io/cypress/63152/workflows/e49a5390-7dc4-432e-abce-391b205c080b/jobs/2615104

```text
  1) src/cy/commands/cookies - no stub
       #getCookies
         returns cookies for the specified domain:

      AssertionError: expected 'key2' to equal 'key1'
      + expected - actual

      -'key2'
      +'key1'

      at  ([native code])
      at  (http://www.barbaz.com:3500/__cypress/tests:1717:41)
  getRet@
  tryCatcher@
  @
  thenFn@
  @
  tryCatcher@
  @
  @
  @
  _drainQueueStep@
  _drainQueue@
  @
  @

  2) src/cy/commands/cookies - no stub
       #getCookie
         returns the cookie from the specified domain:

      AssertionError: expected 'foobar.com value' to equal 'www.foobar.com value'
      + expected - actual

      -'foobar.com value'
      +'www.foobar.com value'

      at  ([native code])
      at  (http://www.barbaz.com:3500/__cypress/tests:1870:38)
  getRet@
  tryCatcher@
  @
  thenFn@
  @
  tryCatcher@
  @
  @
  @
  _drainQueueStep@
  _drainQueue@
  @
  @

  3) src/cy/commands/cookies
       #setCookie
         can set cookies with sameSite:

      AssertionError: expected { Object (name, value, ...) } to have property 'sameSite' of 'no_restriction', but got 'lax'
      + expected - actual

      -'lax'
      +'no_restriction'

      at  ([native code])
  applyChainer@
  arrayReduce@
  reduce@
  applyChainers@
  tryCatcher@
  @
  shouldFn@
  @
  tryCatcher@
  @
  @
  @
  _drainQueueStep@
  _drainQueue@
  @
  @
```

### Other

- See [WebKit (Experimental)](https://docs.cypress.io/guides/guides/launching-browsers#WebKit-Experimental)
- https://playwright.dev/docs/browsers#webkit
- See https://github.com/cypress-io/cypress/pull/29955 for failed attempt to update.

Further test results which may assist troubleshooting:

Results of `yarn workspace @packages/driver cypress:run --browser webkit --spec cypress/e2e/commands/cookies.cy.js` dependent on version of [playwright-webkit](https://www.npmjs.com/package/playwright-webkit)

| Playwright | pass | fail | WebKit | Notes |
| ---------- | ---- | ---- | ------ | :---: |
| 1.24.2     | 129  | 0    | 16.0   |  (1)  |
| 1.34.3     | 129  | 0    | 16.4   |  (2)  |
| 1.35.0     | 129  | 0    | 16.4   |  (3)  |
| 1.36.2     | 129  | 0    | 17.0   |  (3)  |
| 1.37.0     | 128  | 1    | 17.0   |  (4)  |
| 1.37.1     | 127  | 2    | 17.0   |  (5)  |
| 1.38.1     | 127  | 2    | 17.0   |  (5)  |
| 1.39.0     | 127  | 2    | 17.4   |  (5)  |
| 1.40.1     | 128  | 1    | 17.4   |  (4)  |
| 1.41.2     | 127  | 2    | 17.4   |  (5)  |
| 1.42.1     | 127  | 2    | 17.4   |  (5)  |
| 1.43.1     | 127  | 2    | 17.4   |  (5)  |
| 1.44.1     | 127  | 2    | 17.4   |  (5)  |
| 1.45.0     | 127  | 2    | 17.4   |  (6)  |
| 1.45.1     | 126  | 3    | 17.4   |  (7)  |
| 1.45.3     | 126  | 3    | 17.4   |  (7)  |
| 1.46.0     | 127  | 3    | 18.0   |  (7)  |

(1) Currently configured version
(2) Previously recommended version
(3) Test does not exit and instead wrongly opens Cypress runner at end in global mode
(4) Failure is
    AssertionError: expected 'foobar.com value' to equal 'www.foobar.com value'
(5) Failures as (4) above, additionally
    AssertionError: expected 'key2' to equal 'key1'
(6) Failure as (4) above, additionally
   AssertionError: expected { Object (name, value, ...) } to have property 'sameSite' of 'no_restriction', but got 'lax'
(7) Failures as (4), (5) & (6) above

Change version with for instance:

```shell
yarn add playwright-webkit@1.45.3 -D -W
yarn playwright install-deps webkit
```

Test with:

```shell
yarn workspace @packages/driver cypress:run --browser webkit --spec cypress/e2e/commands/cookies.cy.js
```

View WebKit version with

```shell
yarn workspace @packages/driver cypress:open --e2e --browser webkit
```
Settings > Project settings > Resolved configuration (Fails to launch Cypress Runner with some old versions)

For debug logs use:

```shell
export DEBUG=cypress:server:browsers:webkit-automation
```

--- a/.circleci/src/pipeline/@pipeline.yml
+++ b/.circleci/src/pipeline/@pipeline.yml
@@ -8,8 +8,8 @@ version: 2.1
 chrome-stable-version: &chrome-stable-version "141.0.7390.107"
 chrome-beta-version: &chrome-beta-version "142.0.7444.34"
 firefox-stable-version: &firefox-stable-version "142.0"
-base-internal-bullseye: &base-internal-bullseye cypress/base-internal:22.19.0-bullseye
-base-internal-yarn-berry: &base-internal-yarn-berry cypress/base-internal:22.19.0-yarn-berry
+base-internal-trixie: &base-internal-trixie cypress/base-internal:22.19.0-trixie
+base-internal-yarn-berry: &base-internal-yarn-berry cypress/base-internal:22.19.0-yarn-berry-trixie
 # Lowest Node.js version we support of the minimum major version supported
 # See https://docs.cypress.io/app/get-started/install-cypress#Nodejs
 base-internal-minimum-node: &base-internal-minimum-node cypress/base-internal:20.15.0
@@ -46,7 +46,7 @@ executors:
   # the Docker image with Cypress dependencies and Chrome browser
   cy-doc:
     docker:
-      - image: *base-internal-bullseye
+      - image: *base-internal-trixie
     # by default, we use "medium" to balance performance + CI costs. bump or reduce on a per-job basis if needed.
     resource_class: medium
     environment:
@@ -54,7 +54,7 @@ executors:
 
   kitchensink-executor:
     docker:
-      - image: *base-internal-bullseye
+      - image: *base-internal-trixie
     # by default, we use "medium" to balance performance + CI costs. bump or reduce on a per-job basis if needed.
     resource_class: medium
     environment:
@@ -63,7 +63,7 @@ executors:
   # Docker image with non-root "node" user
   non-root-docker-user:
     docker:
-      - image: *base-internal-bullseye
+      - image: *base-internal-trixie
         user: node
     environment:
       PLATFORM: linux
@@ -587,7 +587,7 @@ commands:
                       version: << parameters.google-chrome-for-testing-version >>
                       install_chromedriver: false
 
-  # This code builds better-sqlite3 on Debian 11 (Bullseye). This is necessary because Debian 10 has the oldest glibc version (2.31) that we support.
+  # This code builds better-sqlite3 on Debian 11 (Bullseye). This is necessary because Debian 11 has the oldest glibc version (2.31) that we support.
   #
   # Since this is running Docker remote (because the job running the command may not be using an executor with the appropriate glibc version), we need to 
   # copy the project into the container, and copy the built plugin out of the container because the host running docker does not have access to the 
@@ -1937,7 +1937,7 @@ jobs:
     parallelism: 1
     working_directory: ~/cypress
     docker:
-      - image: *base-internal-bullseye
+      - image: *base-internal-trixie
     environment:
       # needed to inform the bootstrap-docker-container.sh script to link the binary in the system-test project directory
       REPO_DIR: /root/cypress

--- a/cli/CHANGELOG.md
+++ b/cli/CHANGELOG.md
@@ -20,6 +20,7 @@ _Released 10/20/2025 (PENDING)_
 - Add top padding for command log labels. Addressed in [#32774](https://github.com/cypress-io/cypress/pull/32774).
 - The hitbox for expanding a grouped command has been widened. Addresses [#32778](https://github.com/cypress-io/cypress/issues/32778). Addressed in [#32783](https://github.com/cypress-io/cypress/pull/32783).
 - Have cursor on hover of the AUT URL to show as pointer. Addresses [#32777](https://github.com/cypress-io/cypress/issues/32777). Addressed in [#32782](https://github.com/cypress-io/cypress/pull/32782).
+- WebKit now prefers a cookie's fully qualified `domain` when requesting a cookie value via [`cy.getCookie()`](https://docs.cypress.io/api/commands/getcookie). If none are found, the cookie's apex domain will be used as a fallback. Addresses [#29954](https://github.com/cypress-io/cypress/issues/29954), [#29973](https://github.com/cypress-io/cypress/issues/29973) and [#30392](https://github.com/cypress-io/cypress/issues/30392). Addressed in [#32852](https://github.com/cypress-io/cypress/pull/32852).
 
 **Dependency Updates:**
 

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -31,7 +31,7 @@ services:
       - .:/opt/cypress
   ci:
     # This should mirror the image used in workflows.yml
-    image: cypress/base-internal:22.19.0-bullseye
+    image: cypress/base-internal:22.19.0-trixie
     ports:
       - 5566:5566
       - 5567:5567

--- a/package.json
+++ b/package.json
@@ -189,7 +189,7 @@
     "npm-packlist": "9.0.0",
     "p-defer": "^3.0.0",
     "patch-package": "8.0.0",
-    "playwright-webkit": "1.24.2",
+    "playwright-webkit": "1.56.1",
     "pluralize": "8.0.0",
     "print-arch": "1.0.0",
     "proxyquire": "2.1.3",

--- a/packages/driver/cypress/e2e/commands/cookies.cy.js
+++ b/packages/driver/cypress/e2e/commands/cookies.cy.js
@@ -123,10 +123,13 @@ describe('src/cy/commands/cookies - no stub', () => {
 
       cy.getCookies({ domain: 'www.foobar.com' }).then((cookies) => {
         expect(cookies).to.have.length(2)
-        expect(cookies[0].name).to.equal('key1')
-        expect(cookies[0].domain).to.match(/\.?www\.foobar\.com/)
-        expect(cookies[1].name).to.equal('key2')
-        expect(cookies[1].domain).to.match(/\.?foobar\.com/)
+
+        const sortedCookies = Cypress._.sortBy(cookies, 'name')
+
+        expect(sortedCookies[0].name).to.equal('key1')
+        expect(sortedCookies[0].domain).to.match(/\.?www\.foobar\.com/)
+        expect(sortedCookies[1].name).to.equal('key2')
+        expect(sortedCookies[1].domain).to.match(/\.?foobar\.com/)
       })
 
       cy.getCookies({ domain: 'barbaz.com' }).then((cookies) => {
@@ -1380,8 +1383,12 @@ describe('src/cy/commands/cookies', () => {
       // other browsers do not return sameSite at all
       const sameSite = (
         Cypress.isBrowser('webkit')
-      ) ? 'no_restriction' :
-        Cypress.isBrowser('firefox') ? 'unspecified' : null
+        // There is an odd inconsistency when setting an undefined cookie context in WebKit in WebKit 26.
+        // On linux based machines, the sameSite will be set to `lax`, but on MacOS the sameSite is set to `None` (no_restriction).
+        // There is no difference in the webkit-automation logic within cypress so the inconsistency must lie in the OS
+        // implement of WebKit
+      ) ? Cypress.platform === 'darwin' ? 'no_restriction' : 'lax'
+        : Cypress.isBrowser('firefox') ? 'unspecified' : null
 
       if (sameSite) {
         cy.getCookie('five').should('include', { sameSite })

--- a/packages/electron/.cursor/rules/electron-upgrade.mdc
+++ b/packages/electron/.cursor/rules/electron-upgrade.mdc
@@ -67,14 +67,14 @@ When Electron is upgraded, the bundled Node.js version often changes. Sync this 
 ```bash
 # File: /docker-compose.yml
 # Update the base-internal image reference:
-# "cypress/base-internal:22.18.0-bullseye" -> "cypress/base-internal:NEW_NODE_VERSION-bullseye"
+# "cypress/base-internal:22.18.0-trixie" -> "cypress/base-internal:NEW_NODE_VERSION-trixie"
 ```
 
 ### CircleCI Workflows
 ```bash
 # File: /.circleci/src/@workflows.yml
 # Update base-internal image references:
-# "base-internal-bullseye: &base-internal-bullseye cypress/base-internal:22.18.0-bullseye" -> "base-internal-bullseye: &base-internal-bullseye cypress/base-internal:NEW_NODE_VERSION-bullseye"
+# "base-internal-trixie: &base-internal-trixie cypress/base-internal:22.18.0-trixie" -> "base-internal-trixie: &base-internal-trixie cypress/base-internal:NEW_NODE_VERSION-trixie"
 # "base-internal-yarn-berry: &base-internal-yarn-berry cypress/base-internal:22.18.0-yarn-berry" -> "base-internal-yarn-berry: &base-internal-yarn-berry cypress/base-internal:NEW_NODE_VERSION-yarn-berry"
 # "base-internal-minimum-node: &base-internal-minimum-node cypress/base-internal:22.18.0" -> "base-internal-minimum-node: &base-internal-minimum-node cypress/base-internal:NEW_NODE_VERSION"
 ```

--- a/packages/server/lib/automation/util.ts
+++ b/packages/server/lib/automation/util.ts
@@ -12,8 +12,9 @@ export type CyCookie = Pick<chrome.cookies.Cookie, 'name' | 'value' | 'expiratio
 // @ts-ignore
 export type CyCookieFilter = chrome.cookies.GetAllDetails
 
-export const cookieMatches = (cookie: CyCookie | playwright.Cookie, filter?: CyCookieFilter) => {
-  if (filter?.domain && !domainMatch(filter?.domain, cookie.domain)) {
+export const cookieMatches = (cookie: CyCookie | playwright.Cookie, filter?: CyCookieFilter, options?: { strictDomain: boolean }) => {
+  if (filter?.domain) {
+    if (options?.strictDomain ? filter?.domain !== cookie.domain : !domainMatch(filter?.domain, cookie.domain))
     return false
   }
 

--- a/packages/server/lib/browsers/webkit-automation.ts
+++ b/packages/server/lib/browsers/webkit-automation.ts
@@ -270,11 +270,19 @@ export class WebKitAutomation {
 
     if (!cookies.length) return null
 
-    const cookie = cookies.find((cookie) => {
-      return cookieMatches(cookie, filter)
+    // first attempt to match cookie on strict domain
+    let cookie = cookies.find((cookie) => {
+      return cookieMatches(cookie, filter, { strictDomain: true })
     })
 
-    if (!cookie) return null
+    if (!cookie) {
+      cookie = cookies.find((cookie) => {
+          // if unable to match closest via strict domain, then return a cookie that matches the apex domain
+        return cookieMatches(cookie, filter)
+      })
+
+      if (!cookie) return null
+    }
 
     return normalizeGetCookieProps(cookie)
   }

--- a/packages/server/package.json
+++ b/packages/server/package.json
@@ -198,7 +198,7 @@
     "mocked-env": "1.2.4",
     "mockery": "2.1.0",
     "nock": "13.2.9",
-    "playwright-webkit": "1.24.2",
+    "playwright-webkit": "1.56.1",
     "proxyquire": "2.1.3",
     "repl.history": "0.1.4",
     "request-promise": "4.2.6",

--- a/packages/server/test/unit/automation/util_spec.ts
+++ b/packages/server/test/unit/automation/util_spec.ts
@@ -45,5 +45,19 @@ context('lib/automation/util', () => {
 
       expect(cookieMatches(cookie, filter)).to.be.false
     })
+
+    it('strict matches exact domain with strictDomain=true', () => {
+      const cookie = { domain: 'www.example.com' } as CyCookie
+      const filter = { domain: 'www.example.com' }
+
+      expect(cookieMatches(cookie, filter, { strictDomain: true })).to.be.true
+    })
+
+    it('fails apex domain match with strictDomain=true', () => {
+      const cookie = { domain: 'example.com' } as CyCookie
+      const filter = { domain: 'www.example.com' }
+
+      expect(cookieMatches(cookie, filter, { strictDomain: true })).to.be.false
+    })
   })
 })

--- a/system-tests/test-binary/module_api_spec.ts
+++ b/system-tests/test-binary/module_api_spec.ts
@@ -3,7 +3,7 @@ import systemTests from '../lib/system-tests'
 describe('module API', () => {
   systemTests.it('can run module API Mocha spec', {
     timeout: 240000,
-    dockerImage: 'cypress/base-internal:22.19.0-bullseye',
+    dockerImage: 'cypress/base-internal:22.19.0-trixie',
     withBinary: true,
     project: 'module-api',
     browser: 'electron',

